src/metal/rt64_metal.cpp

```diff
@@ -0,0 +1,2419 @@
+        assert(data != nullptr);
+        assert(size > 0);
+        assert(format == RenderShaderFormat::METAL);
+
+        this->format = format;
+        this->functionName = (entryPointName != nullptr) ? NS::String::string(entryPointName, NS::UTF8StringEncoding) : MTLSTR("");
+
+        NS::Error *error;
+        dispatch_data_t dispatchData = dispatch_data_create(data, size, dispatch_get_main_queue(), ^{});
+        library = device->mtl->newLibrary(dispatchData, &error);
+
+        if (error != nullptr) {
+            fprintf(stderr, "MTLDevice newLibraryWithSource: failed with error %s.\n", error->localizedDescription()->utf8String());
+            return;
+        }
+    }
+
+    MetalShader::~MetalShader() {
+        functionName->release();
+        library->release();
+    }
+
+    MTL::Function* MetalShader::createFunction(const RenderSpecConstant *specConstants, uint32_t specConstantsCount) const {
+        if (specConstants != nullptr) {
+            MTL::FunctionConstantValues *values = MTL::FunctionConstantValues::alloc()->init();
+            for (uint32_t i = 0; i < specConstantsCount; i++) {
+                const RenderSpecConstant &specConstant = specConstants[i];
+                values->setConstantValue(&specConstant.value, MTL::DataTypeUInt, specConstant.index);
+            }
+
+            NS::Error *error;
+            auto function = library->newFunction(functionName, values, &error);
+//            values->release();
+
+            if (error != nullptr) {
+                fprintf(stderr, "MTLLibrary newFunction: failed with error: %ld.\n", error->code());
+                return nullptr;
+            }
+
+            return function;
+        } else {
+            return library->newFunction(functionName);
+        }
+    }
+
+    // MetalSampler
+
+    MetalSampler::MetalSampler(MetalDevice *device, const RenderSamplerDesc &desc) {
+        assert(device != nullptr);
+
+        MTL::SamplerDescriptor *descriptor = MTL::SamplerDescriptor::alloc()->init();
+        descriptor->setSupportArgumentBuffers(true);
+        descriptor->setMinFilter(metal::mapSamplerMinMagFilter(desc.minFilter));
+        descriptor->setMagFilter(metal::mapSamplerMinMagFilter(desc.magFilter));
+        descriptor->setMipFilter(metal::mapSamplerMipFilter(desc.mipmapMode));
+        descriptor->setSAddressMode(metal::mapSamplerAddressMode(desc.addressU));
+        descriptor->setTAddressMode(metal::mapSamplerAddressMode(desc.addressV));
+        descriptor->setRAddressMode(metal::mapSamplerAddressMode(desc.addressW));
+        descriptor->setMaxAnisotropy(desc.maxAnisotropy);
+        descriptor->setCompareFunction(metal::mapCompareFunction(desc.comparisonFunc));
+        descriptor->setLodMinClamp(desc.minLOD);
+        descriptor->setLodMaxClamp(desc.maxLOD);
+        descriptor->setBorderColor(metal::mapSamplerBorderColor(desc.borderColor));
+
+        this->state = device->mtl->newSamplerState(descriptor);
```

Comment:
Does the sampler state make a copy of the descriptor internally?

Reply:
The "new" prefixed functions have the responsibility of retaining and releasing when they're done using. We can release from our end to lower the resource count.